File: linkis-commons/linkis-storage/src/main/scala/org/apache/linkis/storage/utils/StorageUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.storage.utils

import org.apache.linkis.common.io.{Fs, FsPath}
import org.apache.linkis.common.utils.{Logging, Utils}
import org.apache.linkis.hadoop.common.conf.HadoopConf
import org.apache.linkis.storage.{LineMetaData, LineRecord}
import org.apache.linkis.storage.errorcode.LinkisStorageErrorCodeSummary.CONFIGURATION_NOT_READ
import org.apache.linkis.storage.exception.StorageFatalException
import org.apache.linkis.storage.resultset.{ResultSetFactory, ResultSetReader, ResultSetWriter}

import org.apache.commons.lang3.StringUtils

import java.io.{Closeable, File, InputStream, OutputStream}
import java.lang.reflect.Method
import java.text.NumberFormat

import scala.collection.mutable

object StorageUtils extends Logging {

  val HDFS = "hdfs"
  val FILE = "file"
  val OSS = "oss"
  val S3 = "s3"
  val BLOB = "https"

  val FILE_SCHEMA = "file://"
  val HDFS_SCHEMA = "hdfs://"
  val OSS_SCHEMA = "oss://"
  val S3_SCHEMA = "s3://"
  val BLOB_SCHEMA = "https://"

  private val nf = NumberFormat.getInstance()
  nf.setGroupingUsed(false)
  nf.setMaximumFractionDigits(StorageConfiguration.DOUBLE_FRACTION_LEN.getValue)

  def doubleToString(value: Double): String = {
    if (value.isNaN) {
      "NaN"
    } else {
      nf.format(value)
    }
  }

  def loadClass[T](classStr: String, op: T => String): Map[String, T] = {
    val _classes = classStr.split(",")
    val classes = mutable.LinkedHashMap[String, T]()
    for (clazz <- _classes) {
      Utils.tryAndError {
        val obj = Utils.getClassInstance[T](clazz.trim)
        classes += op(obj) -> obj
      }
    }
    classes.toMap
  }

  /**
   * Get the corresponding class by passing in the subclass and package name(通过传入子类和包名获得对应的class)
   * @param classStr：Class
   *   name(类名)
   * @param pge:Class
   *   package name(类的包名)
   * @param op：Get
   *   key value(获取键值)
   * @tparam T
   * @return
   */
  def loadClasses[T](
      classStr: String,
      pge: String,
      op: Class[T] => String
  ): Map[String, Class[T]] = {
    val _classes: Array[String] =
      if (StringUtils.isEmpty(pge)) classStr.split(",")
      else classStr.split(",").map { value: String => pge + "." + value }
    val classes = mutable.LinkedHashMap[String, Class[T]]()
    for (clazz <- _classes) {
      Utils.tryAndError({
        val _class =
          Thread.currentThread.getContextClassLoader.loadClass(clazz.trim).asInstanceOf[Class[T]]
        classes += op(_class) -> _class
      })
    }
    classes.toMap
  }

  /**
   * Get the suffix of the file name(获得文件名的后缀)
   * @param path
   * @return
   */
  def pathToSuffix(path: String): String = {
    val fileName = new File(path).getName
    if ((fileName != null) && (fileName.length > 0)) {
      val dot: Int = fileName.lastIndexOf('.')
      if ((dot > -1) && (dot < (fileName.length - 1))) return fileName.substring(dot + 1)
    }
    fileName
  }

  /**
   * Reflection calling method(反射调用方法)
   * @param obj
   * @param method
   * @param args
   * @return
   */
  def invoke(obj: Any, method: Method, args: Array[AnyRef]): Any = {
    method.invoke(obj, args)
  }

  /**
   * Serialized string is a result set of type Text(序列化字符串为Text类型的结果集)
   * @param value
   * @return
   */
  def serializerStringToResult(value: String): String = {
    val resultSet = ResultSetFactory.getInstance.getResultSetByType(ResultSetFactory.TEXT_TYPE)
    val writer = ResultSetWriter.getResultSetWriter(resultSet, Long.MaxValue, null)
    val metaData = new LineMetaData()
    val record = new LineRecord(value)
    writer.addMetaData(metaData)
    writer.addRecord(record)
    val res = writer.toString()
    Utils.tryQuietly(writer.close())
    res
  }

  /**
   * The result set of serialized text is a string(序列化text的结果集为字符串)
   * @param result
   * @return
   */
  def deserializerResultToString(result: String): String = {
    val resultSet = ResultSetFactory.getInstance.getResultSetByType(ResultSetFactory.TEXT_TYPE)
    val reader = ResultSetReader.getResultSetReader(resultSet, result)
    reader.getMetaData
    val sb = new StringBuilder
    while (reader.hasNext) {
      val record = reader.getRecord.asInstanceOf[LineRecord]
      sb.append(record.getLine)
    }
    val value = sb.toString()
    Utils.tryQuietly(reader.close())
    value
  }

  def close(outputStream: OutputStream): Unit = {
    close(outputStream, null, null)
  }

  def close(inputStream: InputStream): Unit = {
    close(null, inputStream, null)
  }

  def close(fs: Fs): Unit = {
    close(null, null, fs)
  }

  def close(outputStream: OutputStream, inputStream: InputStream, fs: Fs): Unit = {
    Utils.tryFinally(if (outputStream != null) outputStream.close())()
    Utils.tryFinally(if (inputStream != null) inputStream.close())()
    Utils.tryFinally(if (fs != null) fs.close())()
  }

  def close(closeable: Closeable): Unit = {
    Utils.tryFinally(if (closeable != null) closeable.close())()
  }

  def getJvmUser: String = System.getProperty("user.name")

  def isHDFSNode: Boolean = {
    val confPath = new File(HadoopConf.hadoopConfDir)
    // TODO IO-client mode need return false
    if (!confPath.exists() || confPath.isFile) {
      throw new StorageFatalException(
        CONFIGURATION_NOT_READ.getErrorCode,
        CONFIGURATION_NOT_READ.getErrorDesc
      )
    } else true
  }

  /**
   * Returns the FsPath by determining whether the path is a schema. By default, the FsPath of the
   * file is returned. 通过判断path是否为schema来返回FsPath，默认返回file的FsPath
   * @param path
   * @return
   */
  def getFsPath(path: String): FsPath = {
    if (
        path.startsWith(FILE_SCHEMA) || path.startsWith(HDFS_SCHEMA) || path.startsWith(BLOB_SCHEMA)
    ) new FsPath(path)
    else {
      new FsPath(FILE_SCHEMA + path)
    }
  }

  def readBytes(inputStream: InputStream, bytes: Array[Byte], len: Int): Int = {
    var count = 0
    var readLen = 0
    // 当使用s3存储结果文件时时，com.amazonaws.services.s3.model.S3InputStream无法正确读取.dolphin文件。需要在循环条件添加:
    // readLen >= 0
    // To resolve the issue when using S3 to store result files and
    // com.amazonaws.services.s3.model.S3InputStream to read .dolphin files, you need to add the
    // condition readLen >= 0 in the loop.
    while (readLen < len && readLen >= 0) {
      count = inputStream.read(bytes, readLen, len - readLen)
      if (count == -1 && inputStream.available() < 1) return readLen
      readLen += count
    }
    readLen
  }

  def isIOProxy(): Boolean = {
    StorageConfiguration.ENABLE_IO_PROXY.getValue
  }

  def isHDFSPath(fsPath: FsPath): Boolean = {
    HDFS.equals(fsPath.getFsType)
  }

}

File: linkis-commons/linkis-storage/src/main/java/org/apache/linkis/storage/utils/StorageUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.storage.utils;

import org.apache.linkis.common.io.Fs;
import org.apache.linkis.common.io.FsPath;
import org.apache.linkis.common.io.resultset.ResultSet;
import org.apache.linkis.common.io.resultset.ResultSetReader;
import org.apache.linkis.common.io.resultset.ResultSetWriter;
import org.apache.linkis.common.utils.Utils;
import org.apache.linkis.hadoop.common.conf.HadoopConf;
import org.apache.linkis.storage.LineMetaData;
import org.apache.linkis.storage.LineRecord;
import org.apache.linkis.storage.exception.StorageWarnException;
import org.apache.linkis.storage.resultset.ResultSetFactory;
import org.apache.linkis.storage.resultset.ResultSetReaderFactory;
import org.apache.linkis.storage.resultset.ResultSetWriterFactory;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.curator.utils.CloseableUtils;

import java.io.*;
import java.lang.reflect.Method;
import java.text.NumberFormat;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.apache.linkis.storage.errorcode.LinkisStorageErrorCodeSummary.CONFIGURATION_NOT_READ;

public class StorageUtils {
  private static final Logger logger = LoggerFactory.getLogger(StorageUtils.class);

  public static final String HDFS = "hdfs";
  public static final String FILE = "file";
  public static final String OSS = "oss";
  public static final String S3 = "s3";

  public static final String FILE_SCHEMA = "file://";
  public static final String HDFS_SCHEMA = "hdfs://";
  public static final String OSS_SCHEMA = "oss://";
  public static final String S3_SCHEMA = "s3://";

  private static final NumberFormat nf = NumberFormat.getInstance();

  static {
    nf.setGroupingUsed(false);
    nf.setMaximumFractionDigits((int) StorageConfiguration.DOUBLE_FRACTION_LEN.getValue());
  }

  public static String doubleToString(double value) {
    return nf.format(value);
  }

  public static <T> Map<String, T> loadClass(String classStr, Function<T, String> op) {
    String[] _classes = classStr.split(",");
    LinkedHashMap<String, T> classes = new LinkedHashMap<>();
    for (String clazz : _classes) {
      try {
        T obj = Utils.getClassInstance(clazz.trim());
        classes.put(op.apply(obj), obj);
      } catch (Exception e) {
        logger.warn("StorageUtils loadClass failed", e);
      }
    }
    return classes;
  }

  public static <T> Map<String, Class<T>> loadClasses(
      String classStr, String pge, Function<Class<T>, String> op) {
    String[] _classes =
        StringUtils.isEmpty(pge)
            ? classStr.split(",")
            : Stream.of(StringUtils.split(classStr, ','))
                .map(value -> pge + "." + value)
                .toArray(String[]::new);
    Map<String, Class<T>> classes = new LinkedHashMap<>();
    for (String clazz : _classes) {
      try {
        Class<T> _class =
            (Class<T>) Thread.currentThread().getContextClassLoader().loadClass(clazz.trim());
        classes.put(op.apply(_class), _class);
      } catch (Exception e) {
        logger.warn("StorageUtils loadClasses failed", e);
      }
    }
    return classes;
  }

  public static String pathToSuffix(String path) {
    String fileName = new File(path).getName();
    if (fileName.length() > 0) {
      int dot = fileName.lastIndexOf('.');
      if (dot > -1 && dot < fileName.length() - 1) {
        return fileName.substring(dot + 1);
      }
    }
    return fileName;
  }

  public static Object invoke(Object obj, Method method, Object[] args) throws Exception {
    return method.invoke(obj, args);
  }

  /**
   * Serialized string is a result set of type Text(序列化字符串为Text类型的结果集)
   *
   * @param value
   * @return
   */
  public static String serializerStringToResult(String value) throws IOException {
    ResultSet resultSet =
        ResultSetFactory.getInstance().getResultSetByType(ResultSetFactory.TEXT_TYPE);
    ResultSetWriter writer =
        ResultSetWriterFactory.getResultSetWriter(resultSet, Long.MAX_VALUE, null);
    LineMetaData metaData = new LineMetaData(null);
    LineRecord record = new LineRecord(value);
    writer.addMetaData(metaData);
    writer.addRecord(record);
    String res = writer.toString();
    IOUtils.closeQuietly(writer);
    return res;
  }

  /**
   * The result set of serialized text is a string(序列化text的结果集为字符串)
   *
   * @param result
   * @return
   */
  public static String deserializerResultToString(String result) throws IOException {
    ResultSet resultSet =
        ResultSetFactory.getInstance().getResultSetByType(ResultSetFactory.TEXT_TYPE);
    ResultSetReader reader = ResultSetReaderFactory.getResultSetReader(resultSet, result);
    StringBuilder sb = new StringBuilder();
    while (reader.hasNext()) {
      LineRecord record = (LineRecord) reader.getRecord();
      sb.append(record.getLine());
    }
    reader.close();
    return sb.toString();
  }

  public static void close(OutputStream outputStream) {
    close(outputStream, null, null);
  }

  public static void close(InputStream inputStream) {
    close(null, inputStream, null);
  }

  public static void close(Fs fs) {
    close(null, null, fs);
  }

  public static void close(OutputStream outputStream, InputStream inputStream, Fs fs) {
    try {
      if (outputStream != null) outputStream.close();
    } catch (IOException e) {
      // ignore exception
    }
    try {
      if (inputStream != null) inputStream.close();
    } catch (IOException e) {
      // ignore exception
    }
    try {
      if (fs != null) fs.close();
    } catch (IOException e) {
      // ignore exception
    }
  }

  public static void close(Closeable closeable) {
    CloseableUtils.closeQuietly(closeable);
  }

  public static String getJvmUser() {
    return System.getProperty("user.name");
  }

  public static boolean isHDFSNode() {
    File confPath = new File(HadoopConf.hadoopConfDir());
    // TODO IO-client mode need return false
    if (!confPath.exists() || confPath.isFile()) {
      throw new StorageWarnException(
          CONFIGURATION_NOT_READ.getErrorCode(), CONFIGURATION_NOT_READ.getErrorDesc());
    } else return true;
  }

  /**
   * Returns the FsPath by determining whether the path is a schema. By default, the FsPath of the
   * file is returned.
   *
   * @param path
   * @return
   */
  public static FsPath getFsPath(String path) {
    if (path.startsWith(FILE_SCHEMA) || path.startsWith(HDFS_SCHEMA)) {
      return new FsPath(path);
    } else {
      return new FsPath(FILE_SCHEMA + path);
    }
  }

  public static int readBytes(InputStream inputStream, byte[] bytes, int len) {
    int readLen = 0;
    try {
      int count = 0;
      // 当使用s3存储结果文件时时，com.amazonaws.services.s3.model.S3InputStream无法正确读取.dolphin文件。需要在循环条件添加:
      // readLen >= 0
      // To resolve the issue when using S3 to store result files and
      // com.amazonaws.services.s3.model.S3InputStream to read .dolphin files, you need to add the
      // condition readLen >= 0 in the loop.
      while (readLen < len && readLen >= 0) {
        count = inputStream.read(bytes, readLen, len - readLen);

        if (count == -1 && inputStream.available() < 1) {
          return readLen;
        }
        readLen += count;
      }
    } catch (IOException e) {
      logger.warn("FileSystemUtils readBytes failed", e);
    }
    return readLen;
  }

  public static String colToString(Object col, String nullValue) {
    if (col == null) {
      return nullValue;
    } else if (col instanceof Double) {
      return doubleToString((Double) col);
    } else if ("NULL".equals(col) || "".equals(col)) {
      return nullValue;
    } else {
      return col.toString();
    }
  }

  public static String colToString(Object col) {
    return colToString(col, "NULL");
  }

  public static boolean isIOProxy() {
    return (boolean) StorageConfiguration.ENABLE_IO_PROXY.getValue();
  }

  public static byte[] mergeByteArrays(byte[] arr1, byte[] arr2) {
    byte[] mergedArray = new byte[arr1.length + arr2.length];
    System.arraycopy(arr1, 0, mergedArray, 0, arr1.length);
    System.arraycopy(arr2, 0, mergedArray, arr1.length, arr2.length);
    return mergedArray;
  }
}
